source/opt/folding_rules.cpp

```diff
@@ -42,6 +42,100 @@ FoldingRule IntMultipleBy1() {
   };
 }
 
+FoldingRule CompositeConstructFeedingExtract() {
+  return [](ir::Instruction* inst,
+            const std::vector<const analysis::Constant*>&) {
+    // If the input to an OpCompositeExtract is an OpCompositeConstruct,
+    // then we can simply use the appropriate element in the construction.
+    assert(inst->opcode() == SpvOpCompositeExtract &&
+           "Wrong opcode.  Should be OpCompositeExtract.");
+    analysis::DefUseManager* def_use_mgr = inst->context()->get_def_use_mgr();
+    uint32_t cid = inst->GetSingleWordInOperand(kExtractCompositeIdInIdx);
+    ir::Instruction* cinst = def_use_mgr->GetDef(cid);
+
+    if (cinst->opcode() != SpvOpCompositeConstruct) {
+      return false;
+    }
+
+    // Get the element being extracted from the OpCompositeConstruct
+    uint32_t element_index = inst->GetSingleWordInOperand(1);
+    uint32_t element_id = cinst->GetSingleWordInOperand(element_index);
```

Comment:
OpCompositeConstruct can create a vector out of component vectors (e.g. float4 out of two float2s or a float2 and 2 floats). If the code is:
```
%construct = OpCompositeConstruct %float4 %float2a %float2b
%extract = OpCompositeExtract %float %construct 3
```
Then this rule will have an out-of-bounds access on the operands. I think you need a special case for vectors. This should also be an unit test.

Reply:
Good catch.  I've updated the folding rule to handle this.  I've also added a few tests.